[Experiment] Blocks: Add a new shared Button block

Copons opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Alternative approach to https://github.com/Automattic/jetpack/pull/14920

See #14935

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->

In #14920 I've tried creating a Button component, complete with utility functions, ready to be used inside existing blocks, replacing our many re-implementations of the Core's Button block.

Watching some WPBlockTalk talks and talking with @apeatling made me rethink this approach, and try with `InnerBlocks`, which would have many UX improvements over the Button components.

Unfortunately, we still can't use the Core's Button as-is. Our use cases are different, and we need different features, and more flexibility.

Some features of this new `jetpack/button` block:
- Server-side rendering (default).
- Can be saved in post content (optional).
- Can output as `a`, `input type="submit"`, `button type"submit"`. (If saved in post content, it's forced as `a` because otherwise the WPCOM KSES would strip it).
- Can be extended with more attributes, data-attributes, etc.

Before committing to this approach, we should see if it actually works for all blocks currently using buttons.
Examples: #15377

#### Note for @Automattic/explorers 

Most of this PR is copypasted from #14920.
Namely, the Button block files is pretty much `===` the Button component from the other PR, but with adjusted attribute names (`buttonTextColor` becomes `textColor` as it's clear that it refers to the button now).

The big differences are in how it's used by Revue.
This PR is slightly larger because changing from component to inner block makes the deprecation/migration a bit more verbose (especially in the PHP side).

Minor note: there's currently no fill/outline style variations. We can easily add them in a second moment if we want (and we do: Eventbrite uses them!)

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Create a new post.
* Add a Revue block, and insert a Revue username.
* Make sure the Button inside the Revue block works as expected.
* Save the post and view it.
* Make sure the button looks as expected.
* Try filling the Revue form and submitting it.
* Make sure the form works as expected.
* Deactivate Jetpack and reload the post on the front end.
* Make sure the form has been replaced with a simple link to the Revue profile.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* N/A

Copons committed:
Create a Button block

Copons committed:
Adjust attribute names

Copons committed:
More attribute fixes

Copons committed:
Button SSR

Copons committed:
Use the Button block with Revue

Copons committed:
Fix typo

Copons committed:
Add the Revue deprecation/migration

Copons committed:
Keep SSR compatibility with deprecated Revue

Copons committed:
Spacing fixes

Copons committed:
Add more missing things to the Button SSR

Copons committed:
Fix missing translation

Copons commented:
I've tried this with the Eventbrite block.
It seems to work well, but there are a few relatively minor issues that emerged during my tests:

- `InnerBlocks` doesn't work with [`BlockStylesSelector`](https://github.com/Automattic/jetpack/blob/17601a541de4c3cd56a225a29d8876845e9e4824/extensions/shared/components/block-styles-selector/index.js), probably because of an unreported infinite loop.
To make it work, I had to render `InnerBlocks` behind a `! attributes.__isBlockPreview` condition, which means that, in the style selector, the "button" style preview is empty.

- Typically, our embed blocks can either be an embedded iframe, or just a button.
When a block contains **only** an `InnerBlocks` with `Button`, it's extremely hard to select the parent block: clicking around always end up selecting the inner `Button`, and I basically always had to use the block navigation tool.

- In some cases, the inner `Button` uses some of the parent block's attributes.
In Eventbrite, for example, it needs both the `url` (to build the link href) and the `eventId` (for the widget ID).
As far as I can see, we can pass variable values to the `InnerBlocks` template, but only on mount (aka: changes won't propagate to the inner blocks).
In my tests this might not be a big deal: the Eventbrite block sets both `url` and `eventId` before transitioning from placeholder to "actual", and so `InnerBlocks` mounts with the correct values; and also when editing an existing post, `InnerBlocks` will mount with the existing values stored in the post content.
I wonder if we might end up in a situation where we modify those values and `InnerBlocks` keeps using some old values, which would be a very hard to spot issue.
Anybody knows if there's a way to keep `InnerBlocks` in sync with the parent attributes?

cc @mtias that might have some answers!

Copons committed:
Display the block inline to avoid filling the entire parent block clickable area

Copons commented:
Ok! I think I've addressed 2 out of 3 issues!
cc @pablinos @apeatling 

### `InnerBlocks` breaks `BlockStylesSelector`

Turns out, @pablinos had the right hunch in https://github.com/Automattic/jetpack/pull/15377#discussion_r405798972, and the issue had something to do with the block example.
Specifically, not the `Button` block's example, but the example of the parent block!

If the parent block contains `InnerBlocks`, we need to specify them in the example too.
Here's the fix in the Eventbrite example: a8fe2c4

### Impossible to select the parent block when Button is the only block

This was because the Button block is rendered as `display: block` and without margins.
When Button is the only block, it completely fills the parent block's clickable area.

The workaround I'm proposing in bd15da3 simply sets it to `display: inline-block`.
This way, the button element is the only clickable area of the Button block, and clicking on all the remaining whitespace would select the parent block.

| | Before | After |
| - | - | - |
| Click on Button | <img width="638" alt="Screenshot 2020-04-09 at 17 10 18" src="https://user-images.githubusercontent.com/2070010/78916685-c3a79400-7a85-11ea-9d77-994173355543.png"> | <img width="631" alt="Screenshot 2020-04-09 at 17 10 12" src="https://user-images.githubusercontent.com/2070010/78916692-c6a28480-7a85-11ea-97b0-bd2f5dacdb56.png"> |
| Click outside Button | <img width="639" alt="Screenshot 2020-04-09 at 17 09 37" src="https://user-images.githubusercontent.com/2070010/78916700-cacea200-7a85-11ea-99e0-016ef6a5e3f2.png"> | <img width="636" alt="Screenshot 2020-04-09 at 17 09 29" src="https://user-images.githubusercontent.com/2070010/78916710-cdc99280-7a85-11ea-9a4b-bedde2c2c0da.png"> |

This is not a perfect solution, but it definitely feels better.
We could add some vertical margin to make sure that there's always at least some clickable area for the parent block, but I don't have any strong opinions about it.

apeatling commented:
> Anybody knows if there's a way to keep InnerBlocks in sync with the parent attributes?

@Copons I implemented this in my contact form block updates in https://github.com/Automattic/jetpack/pull/15362/commits/48d144da1c3b3e670bd0dbf685b1feb162cbe58b#

I select the parent block in the child block here:
https://github.com/Automattic/jetpack/pull/15362/commits/48d144da1c3b3e670bd0dbf685b1feb162cbe58b#diff-161c61664862886e8b8ccd593d7ef3ccR128

Then check for differences between the child block attribute value, and the parent block attribute value:
https://github.com/Automattic/jetpack/pull/15362/commits/48d144da1c3b3e670bd0dbf685b1feb162cbe58b#diff-63e0f976d4a767879d6304eacac73a3cR26

This works, but I'm sure there is a better way, it might give you some ideas though.

apeatling commented on extensions/blocks/button/constants.js:
```diff
@@ -0,0 +1,11 @@
+/**
+ * WordPress dependencies
+ */
+import { __experimentalUseGradient as useGradient } from '@wordpress/block-editor';
```
Nitpick: Should this be capitalized? `UseGradient`

Copons replied:
It shouldn't: `useGradient` is a hook, like `useEffect`.

Copons commented:
> > Anybody knows if there's a way to keep InnerBlocks in sync with the parent attributes?
> 
> @Copons I implemented this in my contact form block updates in #15362

Thanks @apeatling, that's really interesting!

I'm playing around that example, and it's indeed of great value.
My idea right now is to add a `passthroughAttributes` attribute to the Button, where the parent block can map its own attributes keys into the Button attributes, then the Button would "apply" the map, check if the value is different, and update its attributes.

#### Example (based roughly on Eventbrite)

Eventbrite has an `eventId` attribute that needs to be passed through to the Button to build an element ID.
The Button has a `uniqueId` attribute to cover this case (which currently is not kept in sync with the parent).

```jsx
// Parent block

<InnerBlocks template={ [ [
  'jetpack/button',
  {
    passthroughAttributes: {
      uniqueId: 'eventId', // { buttonAttribute: 'parentAttribute' }
    },
  }
] ] } />
```

```jsx
// Button block

withSelect( ( select, { attributes, clientId } ) => {
  const { passthroughAttributes } = attributes;
  // { uniqueId: 'eventId' }

  const { getBlockAttributes, getBlockHierarchyRootClientId } = select( 'core/block-editor' );
  const parentAttributes = getBlockAttributes( getBlockHierarchyRootClientId( clientId ) );

  const mappedAttributes = mapValues( passthroughAttributes, key => parentAttributes[ key ] );
  // { uniqueId: 123456 } (actual parent's `eventId` value)

  // Discard all equals attributes
  const attributesToSync = pickBy( mappedAttributes, ( value, key ) => value !== attributes[ key ] );

  return { attributesToSync };
} )( ButtonEdit );


function ButtonEdit( { attributesToSync, setAttributes } ) {
  useEffect( () => {
    if ( ! isEmpty( attributesToSync ) ) {
      setAttributes( attributesToSync );
    }
  }, [ setAttributes, attributesToSync ] );
}
```

Copons committed:
Avoid ESLint disables

Copons commented:
@Automattic/explorers I think this is already way too big to handle.
I'm proposing to move forward with this approach, which to me feels better and more Gutenberg-y than its component counterpart (#14920).

There are two clear outstanding tasks, but we should take care of them in follow ups:
- Keep the Button attributes in sync with its parent block (see https://github.com/Automattic/jetpack/pull/15370#issuecomment-614071023 for a proposed approach).
- Improve the parent block clickable area when the Button is floating (align left/right).

This is not currently urgent, but reviews are appreciated nonetheless! 🙇‍♂️ 
As a note: most of the code here is copypasted from the component approach, so if you gave that a look, this is pretty much identical. 🙂

scruffian commented:
This is looking good. The only thing I noticed was that the button display in the front end and the editor looks off, but this isn't just a problem with this block, I see it with buttons in all our blocks:

<img width="140" alt="Screenshot 2020-04-16 at 14 42 45" src="https://user-images.githubusercontent.com/275961/79463692-14247180-7ff1-11ea-84b2-bca4b49379fc.png">

This might be an issue with my Jetpack/Gutenberg setup

scruffian reviewed (dismissed):
I think this looks good, lets merge and iterate.

jeherve reviewed (changes requested):
When I enter a custom text for the button content, it does not appear on the frontend.

******

For some reason, this makes me lose all blocks in the editor in WordPress 5.3.2. Anyone else can reproduce?

Copons committed:
Fix incorrect strpos check

Copons committed:
Change Button block category into "layout"

Copons committed:
Fix Revue example preview

Copons commented:
@jeherve 
> When I enter a custom text for the button content, it does not appear on the frontend.

Fixed, it was an incorrect `strpos` check that I was sure I had fixed days ago. Likely forgot to commit it and dropped the change. 🤷 

> For some reason, this makes me lose all blocks in the editor in WordPress 5.3.2. Anyone else can reproduce?

I'm testing with `yarn docker:wp core update --version=5.3.2 --force` and I can't see any noticeable issues. 🤔 
Do you have any additional repro steps I could try?

Also, I've pushed a fix for the Revue example, adding the inner blocks definition in there as well.
It prevented the block preview to show, breaking the entire inserter. Could be related?

Copons commented on extensions/blocks/button/index.js:
```diff
@@ -0,0 +1,29 @@
+export const settings = {
+	title: __( 'Button', 'jetpack' ),
+	icon,
+	category: supportsCollections() ? 'grow' : 'jetpack',
```
Since this block doesn't show up in the inserter, this doesn't matter much.
Though, I've updated this to `layout`, which is the Core Button category.

jeherve commented:
> Do you have any additional repro steps I could try?

I'll send you credentials for the site where I see this, maybe that can help.

Copons commented:
Apparently I had the Guten plugin active when testing on WP 3.5.2, which is why I couldn't repro.
So, the issue here is that this block breaks WP 3.5.2 _without_ the Gutenberg plugin.
I'll try to figure out what goes wrong! 🙂

Copons committed:
Remove import @wordpress/icons as not supported yet

Copons commented:
@jeherve I've rebased this after the Travis fix, and it now builds successfully!
JN still needs to pick up the changes though, but it now works for me locally on WP 3.5.2 and without Gutenberg plugin.
Culprit was an import of the `@wordpress/icons` package which isn't available on older WP versions.

jeherve reviewed (changes requested):
I'm running into other issues now on WP 5.3.2 (without Gutenberg).

When creating a new Revue block or editing a post with an existing Revue block, I get this error:

```
Uncaught (in promise) TypeError: Cannot read property 'attributes' of undefined
    at Te (blocks.min.js?ver=6.7.2:2)
```

Could you take a look?

Copons committed:
Fix typo

apeatling reviewed (dismissed):
Tested this with the Form block in WP latest, WP 5.3.2, and with/without Gutenberg plugin. All worked as expected.

apeatling commented:
@Copons Let's say I wanted to use this block, but I wanted it to support HTML and alignment settings. Would filtering the settings for the block be the best approach, or should there be something built in to change these settings?

Copons commented:
@jeherve There was a typo that caused some malfunctioning in the border radius slider (cbebed0), but I can't repro that.
(And yes, this time I'm definitely on 5.3.2 without Gutenberg 😅)

I've tried adding and editing Revue blocks in these scenarios:
- Post with an old version of Revue (the old version gets migrated fine, and all the relevant attributes are converted to the inner block).
- Empty post.
- Post with a new version of Revue.

Are there other blocks in the post?
If you're using a JN, do you mind sharing the post ID with me?

Copons commented:
> @Copons Let's say I wanted to use this block, but I wanted it to support HTML and alignment settings. Would filtering the settings for the block be the best approach, or should there be something built in to change these settings?

Good question, I'm not sure if it's possible though.
AFAIK blocks settings need to be defined when the block is registered, so I don't see an obvious way to change them when used. 🤔

apeatling commented:
I think it should be possible using this example? https://developer.wordpress.org/block-editor/developers/filters/block-filters/#blocks-registerblocktype

Copons committed:
Move the Button block to production in order to be used by Revue

Copons commented:
Rebased to fix a conflict.

Also moved the Button block from `beta` to `production`.
I understand this is not very proper, but we need it in production to be used by Revue without needing to be on JP Beta.

> I think it should be possible using this example? https://developer.wordpress.org/block-editor/developers/filters/block-filters/#blocks-registerblocktype

@apeatling I'm not familiar with that filter, but you're welcome to try it! As you might be aware, we're a bit swamped this cycle. 😓 

Adding the HTML support might require a larger update, but I think the alignment could be a very nice improvement for the block itself.
It needs to be approached with a bit of care though, what with all the usual CSS box model and float traps, but I don't expect it to be a super daunting task.

This said, I believe we should try to merge this first, because I'm not very keen on adding new stuff to a thousand-lines PR. 🙂

apeatling commented:
@Copons Sorry I wasn't meaning to imply that you should do it. 😀I wanted to make sure that's an option for when this is merged that I can implement. Totally agree no more should go into this PR, I can always open a followup PR after. 👍

scruffian commented on extensions/blocks/button/button-colors-panel.js:
```diff
@@ -0,0 +1,83 @@
+	textColor,
+} ) {
+	const ButtonContrastChecker = (
+		<ContrastChecker
```
This only seems to show when the colors are hovered. Once I move the mouse away it disappears...

scruffian reviewed (approved):
This looks good to me. One small issue that I think we can fix in a follow up PR...

jeherve commented on extensions/blocks/revue/revue.php:
```diff
@@ -105,9 +106,15 @@ class="<?php echo esc_attr( $base_class . 'last-name' ); ?>"
-			echo get_revue_button( $attributes );
+		endif;
+
+		// phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped
+		if ( false !== strpos( $content, 'wp-block-jetpack-revue__fallback' ) ) {
+			echo $content;
+		} else {
+			echo get_deprecated_v1_revue_button( $attributes );
+		}
```
Do you think you could add a comment to explain what's going on here? It took me a minute to understand where that class was coming from.

jeherve reviewed (approved):
This seems to work well, I'd suggest merging this as is.

I found one issue with the button though:

![image](https://user-images.githubusercontent.com/426388/80395427-77e15100-88b3-11ea-81bc-cae9e78a1243.png)

This can be fixed in a follow-up PR though.

Copons commented:
r206525-wpcom

apeatling commented:
@Copons I've found a bug I think, the border radius setting does not seem to stick when the editor is reloaded. It saves the attr value so it's rendered on the front end correctly but not back in the editor.

apeatling commented:
Found a fix for this in https://github.com/Automattic/jetpack/pull/15597

kraftbj commented:
r206618-wpcom (removing a file missed by r206525-wpcom)
